src/NServiceBus.Gateway/Sending/GatewaySendBehavior.cs

```diff
@@ -0,0 +1,117 @@
+namespace NServiceBus.Gateway.Sending
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Extensibility;
+    using Notifications;
+    using ObjectBuilder;
+    using Pipeline;
+    using Receiving;
+    using Routing;
+    using Settings;
+    using Transports;
+
+    class GatewaySendBehavior : SatelliteBehavior
+    {
+        readonly IBuilder builder;
+        readonly IManageReceiveChannels channelManager;
+        readonly IDispatchMessages dispatcher;
+        // ReSharper disable once NotAccessedField.Local
+        readonly GatewayTransaction gatewayTransaction;
+        readonly ReadOnlySettings settings;
+        readonly IMessageNotifier messageNotifier;
+
+        public GatewaySendBehavior(IBuilder builder, IManageReceiveChannels channelManager, IMessageNotifier notifier, IDispatchMessages dispatchMessages, ReadOnlySettings settings, GatewayTransaction transaction)
+        {
+            messageNotifier = notifier;
+            this.settings = settings;
+            gatewayTransaction = transaction;
+            dispatcher = dispatchMessages;
+            this.channelManager = channelManager;
+            this.builder = builder;
+        }
+
+        public string InputAddress { get; set; }
+
+        protected override async Task Terminate(IIncomingPhysicalMessageContext context)
+        {
+            var message = context.Message;
+            var headers = message.Headers;
+            var body = message.Body;
+            
+            var destinationSites = GetDestinationSitesFor(headers);
+
+            //if there is more than 1 destination we break it up into multiple dispatches
+            if (destinationSites.Count > 1)
+            {
+                foreach (var destinationSite in destinationSites)
+                {
+                    await CloneAndSendLocal(body, headers, destinationSite).ConfigureAwait(false);
+                }
+
+                return;
+            }
+
+            var destination = destinationSites.FirstOrDefault();
+
+            if (destination == null)
+            {
+                throw new InvalidOperationException("No destination found for message");
+            }
+
+            await SendToSite(body, headers, destination).ConfigureAwait(false);
+        }
+
+
+        IList<Site> GetDestinationSitesFor(Dictionary<string, string> headers)
+        {
+            return builder.BuildAll<IRouteMessagesToSites>()
```

Comment:
Could we get rid of this by adding a register method to the gateway configuration?